src/v/cloud_storage/cache_service.cc

```diff
@@ -0,0 +1,119 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "cloud_storage/logger.h"
+#include "random/generators.h"
+#include "utils/gate_guard.h"
+#include "vlog.h"
+
+#include <seastar/core/coroutine.hh>
+#include <seastar/core/file.hh>
+#include <seastar/core/fstream.hh>
+#include <seastar/core/future.hh>
+#include <seastar/core/seastar.hh>
+#include <seastar/util/defer.hh>
+
+#include <cloud_storage/cache_service.h>
+
+#include <exception>
+#include <optional>
+
+namespace cloud_storage {
+
+cache::cache(std::filesystem::path cache_dir) noexcept
+  : _cache_dir(std::move(cache_dir)) {}
+
+ss::future<> cache::start() {
+    // todo: start maintenance cycle here
+    return ss::make_ready_future<>();
+}
+
+ss::future<> cache::stop() { return _gate.close(); }
+
+ss::future<std::optional<cache_item>> cache::get(std::filesystem::path key) {
+    gate_guard guard{_gate};
+    ss::file cache_file;
+    try {
+        cache_file = co_await ss::open_file_dma(
+          (_cache_dir / key).native(), ss::open_flags::ro);
+    } catch (...) {
+        vlog(
+          cst_log.debug,
+          "Couldn't open file {} to read: {}",
+          (_cache_dir / key).native(),
+          std::current_exception());
+        co_return std::nullopt;
+    }
+
+    auto data_size = co_await cache_file.size();
+    auto data_stream = ss::make_file_input_stream(cache_file);
+    co_return std::optional(cache_item{std::move(data_stream), data_size});
+}
+
+ss::future<>
+cache::put(std::filesystem::path key, ss::input_stream<char> data) {
+    gate_guard guard{_gate};
+
+    auto filename = (_cache_dir / key).filename();
+    auto tmp_filename = std::filesystem::path(
+      filename.string() + random_generators::gen_alphanum_string(4).data()
+      + ".part");
+    auto dir_path = (_cache_dir / key).remove_filename();
+
+    ss::output_stream<char> out;
+    try {
+        co_await ss::recursive_touch_directory(dir_path.string());
+
+        auto flags = ss::open_flags::rw | ss::open_flags::create
+                     | ss::open_flags::truncate;
+        auto cache_file = co_await ss::open_file_dma(
+          (dir_path / tmp_filename).native(), flags);
+
+        out = co_await ss::make_file_output_stream(cache_file);
+        auto close_out = ss::defer([&out] { out.close().get(); });
+
+        co_await ss::copy(data, out);
+        co_await ss::rename_file(
```

Comment:
We should do a `flush()` before renaming so that internal buffers of output_stream are flushed.